flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/scheduler/BatchSchedulingTest.java

```diff
@@ -0,0 +1,521 @@
+		// but no source has been configured.
+		flink.submitJobAndWait(jobGraph, false, TestingUtils.TESTING_DURATION());
+	}
+
+	@Test(expected = JobExecutionException.class)
+	public void testBatchSchedulingSourceWithInput() throws Exception {
+		// Create the JobGraph
+		JobGraph jobGraph = new JobGraph("testBatchSchedulingSourceWithInput");
+		jobGraph.setScheduleMode(ScheduleMode.BATCH_FROM_SOURCES);
+
+		AbstractJobVertex v1 = new AbstractJobVertex("v1");
+
+		AbstractJobVertex v2 = new AbstractJobVertex("v2");
+		v2.connectNewDataSetAsInput(v1, POINTWISE);
+
+		jobGraph.addVertex(v1);
+		jobGraph.addVertex(v2);
+
+		v2.setAsBatchSource();
+
+		// This should throw an Exception, because the configured source has an input vertex.
+		flink.submitJobAndWait(jobGraph, false, TestingUtils.TESTING_DURATION());
+	}
+
+	/**
+	 * <pre>
+	 *        O verify
+	 *        |
+	 *        . <------------- denotes a pipelined result
+	 *        O union
+	 *  +----´|`----+
+	 *  |     |     |
+	 *  ■     ■     ■ <------- denotes a blocking result
+	 *  O     O     O
+	 * src0  src1  src2
+	 * </pre>
+	 */
+	@Test
+	public void testBatchSchedulingLegWise() throws Exception {
+		// Create the JobGraph
+		JobGraph jobGraph = new JobGraph("testBatchSchedulingLegWise");
+		jobGraph.setScheduleMode(ScheduleMode.BATCH_FROM_SOURCES);
+
+		// Union
+		AbstractJobVertex union = new AbstractJobVertex("union");
+		union.setInvokableClass(UnionForwarder.class);
+		union.setParallelism(PARALLELISM);
+		jobGraph.addVertex(union);
+
+		// Create source vertices
+		AbstractJobVertex[] src = new AbstractJobVertex[3];
+		for (int i = 0; i < src.length; i++) {
+			src[i] = new AbstractJobVertex("src " + i);
+			src[i].setInvokableClass(SourceTask.class);
+			src[i].setParallelism(PARALLELISM);
+			src[i].getConfiguration().setInteger("index", i);
+			jobGraph.addVertex(src[i]);
+
+			// Connect sources to union node
+			union.connectNewDataSetAsInput(src[i], POINTWISE, BLOCKING);
+		}
+
+		// Create verify
+		AbstractJobVertex verify = new AbstractJobVertex("verify");
+		verify.setInvokableClass(VerifyIndexes.class);
+		verify.setParallelism(PARALLELISM);
+		verify.getConfiguration().setInteger("maxIndex", src.length);
+
+		verify.connectNewDataSetAsInput(union, POINTWISE, PIPELINED);
+		jobGraph.addVertex(verify);
+
+		// Slot sharing group for the last pipeline
+		SlotSharingGroup ssg = new SlotSharingGroup(union.getID(), verify.getID());
+		union.setSlotSharingGroup(ssg);
+		verify.setSlotSharingGroup(ssg);
+
+		// - Configure batch scheduling ------------------------------------------------------------
+
+		// src0 is the first to go...
+		src[0].setAsBatchSource();
+
+		src[0].addBatchSuccessors(src[1]); // src0 => src1
+
+		src[1].addBatchSuccessors(src[2]); // src1 => src2
+
+		src[2].addBatchSuccessors(union); // src2 => [union => verify]
+
+		flink.submitJobAndWait(jobGraph, false, TestingUtils.TESTING_DURATION());
```

Comment:
But it assumes that it will fail when something is not right, right? What if, for example, the pipelined execution mode is used and no exception is thrown? Shouldn't a test not rely on the proper functioning of what it tries to test?